Fixes #33209

yj7 opened the pull request.

carltongibson commented on django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:
```diff
@@ -108,6 +128,8 @@
         win.close();
     }
 
+    let popupIndex = 0;
+
```
I think this should live at the top ≈ln6.

yj7 replied:
Will make the change. Also could you help with the linter failing. isort doesnt fail while running the pre-commit hook. What am I missing?

carltongibson replied:
I'll have a look.

carltongibson reviewed (commented):
Hi @yj7 — this looks good. 

Do you think there's any way of protecting users from doing the _wrong thing™_ and closing windows in the wrong order? 

Example: open a chain of pop-ups and then close or Save one of the earlier ones. Later ones then don't function: Popup hangs on closing. 

You have to opt into this error, and this is already quite opt-in, but maybe it's bulletproof-able? What do you think?

yj7 commented:
> Hi @yj7 — this looks good.
> 
> Do you think there's any way of protecting users from doing the _wrong thing™_ and closing windows in the wrong order?
> 
> Example: open a chain of pop-ups and then close or Save one of the earlier ones. Later ones then don't function: Popup hangs on closing.
> 
> You have to opt into this error, and this is already quite opt-in, but maybe it's bulletproof-able? What do you think?

I like this, we can add a alert if the user is closing a popup earlier than expected. The alert will still allow the user to close the popup but at least there will be a warning. Not completely bullet proof though. All we would need code wise is a simple flag indicating later popups are still open.

yj7 commented:
> > Hi @yj7 — this looks good.
> > Do you think there's any way of protecting users from doing the _wrong thing™_ and closing windows in the wrong order?
> > Example: open a chain of pop-ups and then close or Save one of the earlier ones. Later ones then don't function: Popup hangs on closing.
> > You have to opt into this error, and this is already quite opt-in, but maybe it's bulletproof-able? What do you think?
> 
> @carltongibson I like this, we can add an alert if the user is closing a popup earlier than expected. The alert will still allow the user to close the popup but at least there will be a warning. Not completely bullet proof though. All we would need code wise is a simple flag indicating later popups are still open.

@carltongibson 
Did some quick reading up. The alert approach may not work. JavaScript does have a way to show an alert warning the user before closing the window but does not support custom text messages; only a generic "changes may not be saved" is shown, varies from browser to browser. This generic message doesn't really help the user.

Another approach could be to close the child popups when closing the parent popup but that leads to a set of different questions. For e.g. What do with the unsaved values? Will the user see it as an bug/unexpected behavior?

felixxm commented on tests/admin_views/admin.py:
```diff
@@ -22,7 +22,7 @@
     Actor, AdminOrderedAdminMethod, AdminOrderedCallable, AdminOrderedField,
     AdminOrderedModelMethod, Album, Answer, Answer2, Article, BarAccount, Book,
-    Bookmark, Category, Chapter, ChapterXtra1, Child, ChildOfReferer, Choice,
+    Bookmark, Box, Category, Chapter, ChapterXtra1, Child, ChildOfReferer, Choice,
```
Import should be wrapped at 79 chars. Please move `Choice` to the next line:
```python
    Bookmark, Box, Category, Chapter, ChapterXtra1, Child, ChildOfReferer,
    Choice, City, Collector, Color, Color2, ComplexSortedPerson, CoverLetter, 
```

carltongibson commented:
@yj7 I think simple-best-effort would be enough. (This is already an edge-case) 

How would closing the child pop-ups look? If that's straightforward I might be tempted to say go for that. (I think Folks would work out what was happening, and I'm not sure more is worth the complexity)

yj7 commented:
> @yj7 I think simple-best-effort would be enough. (This is already an edge-case)
> 
> How would closing the child pop-ups look? If that's straightforward I might be tempted to say go for that. (I think Folks would work out what was happening, and I'm not sure more is worth the complexity)

Closing popups is very straightforward. Just need to create a small function that closes the child popups on window close. The child windows would be closed and unsaved changes will be lost. I think so this a good enough compromise. Want me to go ahead with this?

carltongibson commented:
> ...very straightforward Want me to go ahead with this?

Yes, I think that sounds right. 

Can you add a matching selenium test for that too please? 

Thanks for the effort!

yj7 commented:
> > ...very straightforward Want me to go ahead with this?
> 
> Yes, I think that sounds right.
> 
> Can you add a matching selenium test for that too please?
> 
> Thanks for the effort!

@carltongibson So I went with a mixed approach. Upon closing the parent popup an alert is shown warning the user. If the user still chooses to close the popup then all the child windows are closed automatically.

carltongibson reviewed (commented):
Hi @yj7 — thanks for the quick update. 

I think we need to drop the `alert()` — testing on Firefox, it doesn't work (it's not shown). The new selenium then fails with a `NoSuchWindowException`. 

There are two cases, shutting an ancestor window: 

* I switch to it, and hit save. 
* I switch to it, and close it without saving. 

AFAICS, both of these now behave as expected: in the first, the object is created, and the _child_ popups are shut; in the second, no object is created and the child popups are shut. 

Given the way the popups are layered, TBH, I think this is clear enough, and the alert is unnecessary.

yj7 commented:
> AFAICS, both of these now behave as expected: in the first, the object is created, and the _child_ popups are shut; in the second, no object is created and the child popups are shut.

@carltongibson In the case wherein a user switches to an ancestor popup and hits save. If the child popup has unsaved data; then that data will be lost. For this scenario I added the warning message.

As far as the firefox failure goes, I tried recreating the scenario in the browser and the alert is shown. So it must some issue with how selenium interacts with the browser. Firefox needs some sort of interaction with the window otherwise the JavaScript event is not fired.

I feel the alert is necessary to warn the user about the data loss. Still let me know, i'll remove the alert.

carltongibson commented:
@yj7 Let me have another play. I don't think I tried that exact flow.

carltongibson commented:
buildbot, test on selenium.

carltongibson commented:
Hi @yj7 — 

Please see the failure here: https://djangoci.com/job/pull-requests-selenium/database=sqlite3,label=bionic-pr,python=python3.9/300/testReport/junit/admin_views.tests/SeleniumTests/test_related_popup_index/ — this is what I see locally too. 

>  In the case wherein a user switches to an ancestor popup and hits save & if the child popup has unsaved data; then that data will be lost. For this scenario I added the warning message.

OK, this just isn't working locally. I enter unsaved data, switch, and process a parent without any warning. 

**However** I think that's overkill TBH. I have to really try and do the wrong thing there. Experimenting, I feel that just closing the child popups is sufficient. (Remembering that this is already quite an edge-case, and the _natural behaviour_ here is _right_, I think.)

I think we can strip out the alert and adjust the test to make sure the child popups are close when saving or closing a parent.

yj7 commented:
> Hi @yj7 —
> 
> Please see the failure here: https://djangoci.com/job/pull-requests-selenium/database=sqlite3,label=bionic-pr,python=python3.9/300/testReport/junit/admin_views.tests/SeleniumTests/test_related_popup_index/ — this is what I see locally too.
> 
> > In the case wherein a user switches to an ancestor popup and hits save & if the child popup has unsaved data; then that data will be lost. For this scenario I added the warning message.
> 
> OK, this just isn't working locally. I enter unsaved data, switch, and process a parent without any warning.
> 
> **However** I think an alert here is overkill TBH. I have to really try and do the wrong thing there. Experimenting, I feel that just closing the child popups is sufficient. (Remembering that this is already quite an edge-case, and the _natural behaviour_ here is _right_, I think.)
> 
> I think we can strip out the alert and adjust the test to make sure the child popups are close when saving or closing a parent.

@carltongibson 

Removed the alert and updated the test case.

felixxm commented on django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:
```diff
@@ -4,14 +4,46 @@
+        if(document.getElementsByName("_popup").length > 0) {
+            const index = window.name.lastIndexOf("__") + 2;
+            popupIndex = parseInt(window.name.substring(index));   
+        }
+        else {
```
```suggestion
        } else {
```

carltongibson commented:
Hi @yj7 — this is looking good. 

Are you able to squash, and rebase and adjust the [commit message format](https://docs.djangoproject.com/en/3.2/internals/contributing/committing-code/#committing-guidelines)? — If not I can do it. 👍

yj7 commented:
> Hi @yj7 — this is looking good.
> 
> Are you able to squash, and rebase and adjust the [commit message format](https://docs.djangoproject.com/en/3.2/internals/contributing/committing-code/#committing-guidelines)? — If not I can do it. 👍

@carltongibson 

I tried squashing and rebasing following this [guide](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/#after-review). Please check and let me know.

carltongibson commented:
buildbot, test on selenium.

carltongibson reviewed (approved):
Hi @yj7 — I've squashed down to a single commit and rebased on `main` in https://github.com/django/django/pull/14823/commits/492ed60f236d770eb9a6d56d85ff2550bb1ecfff. 

We'll just wait for CI, but looks good. Thanks for the input! Welcome aboard! ⛵

carltongibson commented:
buildbot, test on selenium.

yj7 commented:
@carltongibson & @felixxm - thanks for the review and feedback. Feels good to be onboard!

felixxm commented:
This PR caused tests failures in the `headless` mode:
- `admin_views.tests.SeleniumTests.test_related_popup_index`
- `admin_views.tests.SeleniumTests.test_search_input_filtered_page`

see [logs](https://djangoci.com/job/django-selenium/887/).

yj7 commented:
> This PR caused tests failures in the `headless` mode:
> 
> * `admin_views.tests.SeleniumTests.test_related_popup_index`
> * `admin_views.tests.SeleniumTests.test_search_input_filtered_page`
> 
> see [logs](https://djangoci.com/job/django-selenium/887/).

@felixxm 

Not sure where the issue is. I ran the tests multiple times locally. Received the error only once, the other times the tests were completed successfully. Looking into it, have a hunch that closing the child windows is causing selenium to misbehave.

smithdc1 commented:
Previously we've seen failures due to the width of the screen in headless mode. See:

https://github.com/django/django/pull/14018

Not sure if that's the same here, but thought I'd mention it.

yj7 commented:
buildbot, test on selenium.

yj7 commented:
> > This PR caused tests failures in the `headless` mode:
> > 
> > * `admin_views.tests.SeleniumTests.test_related_popup_index`
> > * `admin_views.tests.SeleniumTests.test_search_input_filtered_page`
> > 
> > see [logs](https://djangoci.com/job/django-selenium/887/).
> 
> @felixxm
> 
> Not sure where the issue is. Upon running the tests multiple times locally; I received the error only once, the other times the tests were completed successfully. Looking into it, have a hunch that closing the child windows is causing selenium to misbehave.

I received the error without using --headless option also in selenium. Not able to reproduce it every time. Based on the line of error `self.selenium.switch_to.window(self.selenium.window_handles[-1])` and the error message, it seems that `self.selenium.window_handles[-1]` points to the current popup window which has not yet closed, but is in the process of closing. This is also supported by the fact the error is sporadic and is most likely a race condition. To solve this instead of using `self.selenium.window_handles[-1]` we can directly use the respective window handles. Again since I am not able to reliably recreate the issue, there is no way to for me to be sure that this fix is correct. Open to suggestions / ideas.

carltongibson commented:
@yj7 I'm thinking of waiting for the window count to be known (opposite to what we do when opening them). (Not yet sure that'll work but...)

yj7 commented:
> @yj7 I'm thinking of waiting for the window count to be known (opposite to what we do when opening them). (Not yet sure that'll work but...)

@carltongibson - Waiting for the window count would also work. Any preference between the two fixes i.e. waiting for the window count or switching using the window handle? But both these fixes are dependent on the root cause of the error and no way to be sure that the fix solves the issue.

carltongibson commented:
@yj7 Push one, and let's see 😀 (I don't mind which we try first.) Thanks!

yj7 commented:
> Hi @yj7 — I've squashed down to a single commit and rebased on `main` in [492ed60](https://github.com/django/django/commit/492ed60f236d770eb9a6d56d85ff2550bb1ecfff).
> 
> We'll just wait for CI, but looks good. Thanks for the input! Welcome aboard! ⛵

Hi @carltongibson, I am not sure which version of Django would contain these changes. Could you explain?

carltongibson commented:
@yj7 This will be part of Django 4.1. (`main` now will be branched into `stable/4.1.x` for the [alpha release in May 2022](https://code.djangoproject.com/wiki/Version4.1Roadmap)
